File: frontend/src/data/phrasebook.ts
export type { CEFRLevel, Register } from '@/data/types'

import { apiFetch } from '@/lib/api'

export interface Phrase {
  text: string
  context: string
  register: string
  unit_ref?: string
  romanization?: string
}

export interface PhrasebookCategory {
  id: string
  level: string
  situation: string
  icon: string
  phrases: Phrase[]
}

export interface PhrasebookNativeHelpPhraseNote {
  phrase: string
  note: string
}

export interface PhrasebookNativeHelpTrap {
  mistake: string
  fix: string
}

export interface PhrasebookNativeHelpGlossaryItem {
  term: string
  meaning: string
  note?: string
}

export interface PhrasebookNativeHelp {
  summary: string
  usage_tips: string[]
  register_notes: string[]
  phrase_notes: PhrasebookNativeHelpPhraseNote[]
  common_traps: PhrasebookNativeHelpTrap[]
  mini_glossary: PhrasebookNativeHelpGlossaryItem[]
}

export async function getPhrasebookCategories(
  targetLanguage: string = 'en-GB'
): Promise<PhrasebookCategory[]> {
  const res = await apiFetch(
    `/api/phrasebook?language=${encodeURIComponent(targetLanguage)}`
  )
  if (!res.ok) return []
  const data = await res.json()
  return data.categories ?? []
}

export async function getPhrasebookByLevel(
  level: string,
  targetLanguage: string = 'en-GB'
): Promise<PhrasebookCategory[]> {
  const res = await apiFetch(
    `/api/phrasebook/level/${encodeURIComponent(level)}?language=${encodeURIComponent(targetLanguage)}`
  )
  if (!res.ok) return []
  const data = await res.json()
  return data.categories ?? []
}

export async function getPhrasebookNativeHelp(
  categoryId: string,
  targetLanguage: string = 'en-GB'
): Promise<PhrasebookNativeHelp | null> {
  const res = await apiFetch(
    `/api/phrasebook/${encodeURIComponent(categoryId)}/native-help?language=${encodeURIComponent(targetLanguage)}`,
    { method: 'POST' }
  )
  if (!res.ok) return null
  const data = await res.json()
  return data.native_help ?? null
}

File: backend/app/schemas/phrasebook.py
from __future__ import annotations

import warnings

from pydantic import BaseModel

warnings.filterwarnings(
    "ignore",
    message=r'Field name "register" in "PhrasebookEntryResponse" shadows.*',
    category=UserWarning,
)


class PhrasebookEntryResponse(BaseModel):
    text: str
    context: str
    register: str
    unit_ref: str | None = None
    romanization: str | None = None


class PhrasebookCategoryResponse(BaseModel):
    id: str
    level: str
    situation: str
    icon: str
    phrases: list[PhrasebookEntryResponse]


class PhrasebookCategoriesResponse(BaseModel):
    categories: list[PhrasebookCategoryResponse]


class PhrasebookCategoryDetailResponse(BaseModel):
    category: PhrasebookCategoryResponse


class PhrasebookNativeHelpPhraseNoteResponse(BaseModel):
    phrase: str
    note: str


class PhrasebookNativeHelpTrapResponse(BaseModel):
    mistake: str
    fix: str


class PhrasebookNativeHelpGlossaryResponse(BaseModel):
    term: str
    meaning: str
    note: str | None = None


class PhrasebookNativeHelpContentResponse(BaseModel):
    summary: str
    usage_tips: list[str]
    register_notes: list[str]
    phrase_notes: list[PhrasebookNativeHelpPhraseNoteResponse]
    common_traps: list[PhrasebookNativeHelpTrapResponse]
    mini_glossary: list[PhrasebookNativeHelpGlossaryResponse]


class PhrasebookNativeHelpResponse(BaseModel):
    native_help: PhrasebookNativeHelpContentResponse
